transport-native-io_uring/src/main/java/io/netty/channel/uring/IOUringSubmissionQueue.java

```diff
@@ -0,0 +1,286 @@
+    private static final int SQE_RW_FLAGS_FIELD = 28;
+    private static final int SQE_USER_DATA_FIELD = 32;
+    private static final int SQE_PAD_FIELD = 40;
+
+    private static final int KERNEL_TIMESPEC_TV_SEC_FIELD = 0;
+    private static final int KERNEL_TIMESPEC_TV_NSEC_FIELD = 8;
+
+    //these unsigned integer pointers(shared with the kernel) will be changed by the kernel
+    private final long kHeadAddress;
+    private final long kTailAddress;
+    private final long kRingMaskAddress;
+    private final long kRingEntriesAddress;
+    private final long fFlagsAdress;
+    private final long kDroppedAddress;
+    private final long arrayAddress;
+
+    private final long submissionQueueArrayAddress;
+
+    private long sqeHead;
+    private long sqeTail;
+
+    private final int ringSize;
+    private final long ringAddress;
+    private final int ringFd;
+
+    private static final int SOCK_NONBLOCK = 2048;
+    private static final int SOCK_CLOEXEC = 524288;
+
+    private final ByteBuffer timeoutMemory;
+    private final long timeoutMemoryAddress;
+
+    IOUringSubmissionQueue(long kHeadAddress, long kTailAddress, long kRingMaskAddress, long kRingEntriesAddress,
+                                  long fFlagsAdress, long kDroppedAddress, long arrayAddress,
+                                  long submissionQueueArrayAddress, int ringSize,
+                                  long ringAddress, int ringFd) {
+        this.kHeadAddress = kHeadAddress;
+        this.kTailAddress = kTailAddress;
+        this.kRingMaskAddress = kRingMaskAddress;
+        this.kRingEntriesAddress = kRingEntriesAddress;
+        this.fFlagsAdress = fFlagsAdress;
+        this.kDroppedAddress = kDroppedAddress;
+        this.arrayAddress = arrayAddress;
+        this.submissionQueueArrayAddress = submissionQueueArrayAddress;
+        this.ringSize = ringSize;
+        this.ringAddress = ringAddress;
+        this.ringFd = ringFd;
+
+        timeoutMemory = Buffer.allocateDirectWithNativeOrder(KERNEL_TIMESPEC_SIZE);
+        timeoutMemoryAddress = Buffer.memoryAddress(timeoutMemory);
+    }
+
+    public long getSqe() {
+        long next = sqeTail + 1;
+        long kRingEntries = toUnsignedLong(PlatformDependent.getInt(kRingEntriesAddress));
+        long sqe = 0;
+        if ((next - sqeHead) <= kRingEntries) {
+            long index = sqeTail & toUnsignedLong(PlatformDependent.getInt(kRingMaskAddress));
+            sqe = SQE_SIZE * index + submissionQueueArrayAddress;
+            sqeTail = next;
+        }
+        return sqe;
+    }
+
+    private void setData(long sqe, long eventId, EventType type, int fd, long bufferAddress, int length, long offset) {
+        //Todo cleaner
+        //set sqe(submission queue) properties
+        PlatformDependent.putByte(sqe + SQE_OP_CODE_FIELD, (byte) type.getOp());
+        PlatformDependent.putByte(sqe + SQE_FLAGS_FIELD, (byte) 0);
+        PlatformDependent.putShort(sqe + SQE_IOPRIO_FIELD, (short) 0);
```

Comment:
I think you mean SQE right? I don't think the kernel reset all fields to zero, because liburing does the same thing, https://github.com/axboe/liburing/blob/master/src/include/liburing.h#L197 any idea how do that more efficient?

Reply:
I think all sqe entries gonna be reuse, the user application is responsible for that as far as I know

Reply:
Sorry yes meant SQE. Right, IIUC app owns the whole SQE array and reuse strategy for the entries, kernel never writes to it. So we can zero the whole thing up-front and after than don't need to touch fields that we know for our purposes will never be anything but 0 (like the padding). I'm guessing liburing zeros every time since it's more general purpose, it doesn't know that the app will never use certain fields. Not sure why even the padding is zero'd every time but maybe just because it doesn't know for sure what the app is doing with the mem between uses. Or I could still be missing something... :)

Reply:
I agree but I would't zero the whole think up front (like padding) avoid more calls(putLong etc.) only the zero the field we dont use(like padding)
or do you have any idea how to write zeroes only at once?

Reply:
@chrisvest I don't feel super strongly but can't really see how this would be fragile since it's the only place that SQEs are written and the fields that _are_ set are going to be overwritten every time? (and same applies to any fields that might be used in future which aren't now)

Reply:
christ, you mean the whole square entry and not all square entries every time right?

Reply:
I just realized I there may be some misunderstanding re what I was suggesting. By "up front" I mean zero the entire SQ array (all SQE slots) only once when the ring itself is first created (right after `io_uring_setup`). I'm _not_ suggesting to zero each SQE slot each time it's used or to do any conditional setting of any fields.

Reply:
@njhill we could use `memset` for that in JNI, I'm guessing it wouldn't make sense to zero each SQE slot each time, since the kernel read the sqe entry which we submitted